Add the OpenSSF labs process (#421)
ossf/tac: process/labs-process.md

# OpenSSF Labs

<!-- What other names might we consider instead of "OpenSSF Labs"? -->

The OpenSSF Labs provide a space for open source projects that are in the
earliest stages of their lifecycle to experiment, foster collaboration, and grow
their community prior to transitioning into the OpenSSF [project lifecycle].

<!-- Right now, the focus is on Projects, but should we consider broadening the labs to any type of TI? -->

OpenSSF Labs follow a similar model to Hyperledger Labs.

## Benefits

The OpenSSF Labs provide OSS developers several benefits:

* A common governance and legal framework under the OpenSSF that
facilitates cross-organization or -vendor collaboration.
* The lowest barrier to starting brand new projects.
* A dedicated GitHub repository, if starting a lab from scratch.
* A streamlined transition into the [Sandbox stage] of the OpenSSF [project
lifecycle].

> ## Project Responsibilities

Reviewer: ```suggestion
## Lab Responsibilities
```

Developers of OpenSSF labs are responsible for:

* Submitting a [new lab proposal] for review by the [OpenSSF TAC].
* Ensuring all commits are properly signed-off to avoid issues related to [DCO].
* Notifying the TAC if the lab needs to be suspended or archived.

Labs projects are also highly encouraged to engage with the [existing
Technical Initiatives] (working groups, projects or SIGs) in OpenSSF to build
their community and find a potential pathway towards acceptance as an OpenSSF
project.

## Archiving

The TAC will periodically check on the activity of labs. Labs that have been
inactive for an extended period (6+ months), or are explicitly suspended by
the maintainers, will be moved into the [Archived
stage](templates/LAB_NAME_archived_stage.md).

Archived lab repositories are not actively maintained and will be marked as
"archived" (read-only) on GitHub. They can be reactivated if there is interest

> in resuming work on a project.

Reviewer: ```suggestion
in resuming work on a lab.
```

## New Lab Proposal Process

> 1. Fork the `<TBD>` repo.

Reviewer: I would advise having a separate repo for labs. First because it more clearly separates the labs from the other TIs. Indeed, it will be an ongoing battle to keep the 2 concepts lab vs projects separated. Then, the TAC might eventually delegate this task to some other group - in Hyperleger we have a group of "Lab stewards" dealing with this. If it's a separate repo it will be easy to make the change of who's in charge. That's harder to do if it's part of the TAC repo.

2. Fill out the [proposal template](templates/LAB_NAME_lab_stage.md)
   and save it into the `labs` subdirectory under the name of your lab,
   such as `coolnewproject_lab.md`.
   <br/>
> [!TIP]
> It is expected that your lab repository on GitHub will have the same
> name as the proposal, so keep that in mind when submitting your proposal.

3. In the proposal template, there is an entry for sponsor(s). Although this
   is not required, proposers are encouraged to seek a sponsor in the OpenSSF
   community who can help them create ties with the rest of the community
   and review the proposal to make sure it is novel and aligned with the
   [OpenSSF mission].
   <br/>
   To find sponsors:
     1. use your connections to existing projects and ask maintainers,
     2. engage with [existing Technical Initiatives] (working groups, projects

> 	    or SIGs) with affinities to the proposed lab and pitch the project in

Reviewer: ```suggestion
	    or SIGs) with affinities to the proposed lab and pitch it in
```

		their meetings or [Slack channels](https://slack.openssf.org/). It's
		good to have the template already filled out when you reach out.
   <br>
> [!IMPORTANT]

> > Lab sponsors may but are not required to actively participate in

Reviewer: ```suggestion
> Lab sponsors may, but are not required to, actively participate in
```

> the lab once the proposal has been reviewed and accepted.

4. Commit your changes with proper sign-off. This means that your commit
   log message must contain a line that looks like the following one,
   with your actual name and email address:

   `Signed-off-by: John Doe <john.doe@example.com>`

   Adding the `-s` flag to your `git commit` command will add that line
   automatically. You can also add it manually as part of your commit
   log message or add it afterwards with `git commit --amend -s`.

5. Submit a Pull Request to the `<TBD>` repo.

> The [OpenSSF TAC] will then review your proposal. Like sponsors, TAC members may
> but are not required to participate in ongoing work like contributing or

Reviewer: ```suggestion
The [OpenSSF TAC] will then review your proposal. Like sponsors, TAC members may,
but are not required to, participate in ongoing work like contributing or
```

reviewing code in the lab.

> ### Transferring an existing repository

Reviewer 2: Would transferring existing repositories over to the labs org require a full license and IP review as for "full" projects coming into the OpenSSF? That could create a significant overhead and we may want to trade-off the intention of making labs a very accessible path into the OpenSSF space and putting the necessary guardrails in place.

Author: I don't think transferring an existing repo to the labs org should require a full license and IP review, as long as they have one of the expected licenses and DCO. The full license and IP review becomes part of the transition requirement to sandbox stage. I can clarify this in the text.

CC @Naomi-Wash @riaankleinhans for confirmation on this.

Author: I've updated the text with a note about this, please let me know what you think.

By default, OpenSSF staff will create a new GitHub repository for you to
start a new lab in. If you have an existing GitHub repo you would like to
bring to your proposed lab, you have the option to request for that
repo to be transferred into the <TBD GH org> instead.

> However, we require that every commit in the existing repo to br

Reviewer: ```suggestion
However, we require that every commit in the existing repo to bring is
```

signed-off so there are no issues related to [DCO].
If that is not the case, you will need to transfer your existing code by
squashing all of your commits into a single first commit made against
your new lab repo with your sign-off.

### License requirement

> <!-- This is something that other Labs require of the projects they host. What, if any requirements should we include here? -->

Reviewer 2: I suggest that we apply the very same IP requiremnents as defined in the OpenSSF Charter Section 4: 

i) Software source code
(1) Apache License, Version 2.0, available at [https://www.apache.org/licenses/LICENSE- 2.0](https://www.apache.org/licenses/LICENSE-%202.0); or
(2) MIT License available at https://opensource.org/licenses/MIT;
ii) Data
(1) Any of the Community Data License Agreements, available at https://www.cdla.io/;
iii) Specifications
(1) Community Specification License, Version 1.0, available at https://github.com/CommunitySpecification/1.0
iv) All other Documentation
(1) Creative Commons Attribution 4.0 International License, available at https://creative commons.org/licenses/by/4.0/.

Author: done.

of the [OpenSSF charter](https://charter.openssf.org/):

#### Software source code

(1) Apache License, Version 2.0, available at [https://www.apache.org/licenses/LICENSE- 2.0](https://www.apache.org/licenses/LICENSE- 2.0); or

(2) MIT License available at [https://opensource.org/licenses/MIT](https://opensource.org/licenses/MIT)

#### Data

Any of the Community Data License Agreements, available at [https://www.cdla.io](https://www.cdla.io)

#### Specifications

Community Specification License, Version 1.0, available at [https://github.com/CommunitySpecification/1.0](https://github.com/CommunitySpecification/1.0)

#### All other Documentation

(1) Creative Commons Attribution 4.0 International License, available at [https://creative commons.org/licenses/by/4.0/](https://creative commons.org/licenses/by/4.0/)

### Transferring an existing repository

By default, OpenSSF staff will create a new GitHub repository in the ossf-labs
organization for you to start a new lab in. If you have an existing GitHub repo
you would like to bring to your proposed lab, you have the option to request
permission to transfer that repo into the openssf-labs GitHub organization

> instead. This request is noted as part of the [new lab proposal].

Author: @kj-powell can you please double-check this? Are we missing any steps?
